megamek/src/megamek/client/ui/swing/ClientGUI.java

```diff
@@ -2119,6 +2121,79 @@ public void saveListFile(ArrayList<Entity> unitList, String filename) {
         }
     }
 
+    private ProcessBuilder printToMegaMekLab(ArrayList<Entity> unitList, File mmlExecutable, boolean autodetected) {
+        boolean jarfile;
+        try (var ignored = new JarFile(mmlExecutable)) {
+            jarfile = true;
+        } catch (IOException ignored) {
+            jarfile = false;
+        }
+
+        File unitFile;
+        try {
+            unitFile = File.createTempFile("MegaMekPrint", ".mul");
+            EntityListFile.saveTo(unitFile, unitList);
+            unitFile.deleteOnExit();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        String[] command;
+
+        if (!jarfile) {
+            if (!mmlExecutable.canExecute()) {
+                if (autodetected) {
+                    logger.error("Could not auto-detect MegaMekLab! Please configure the path to the MegaMekLab executable in the settings.", "Error printing unit list");
+                } else {
+                    logger.error("%s does not appear to be an executable! You may need to set execute permission or configure the path to the MegaMekLab executable in the settings.".formatted(mmlExecutable.getName()), "Error printing unit list");
+                }
+                return null;
+            }
+
+            if (mmlExecutable.getName().toLowerCase().contains("gradle")) {
+                // If the executable is `gradlew`/`gradelw.bat`, assume it's the gradle wrapper
+                // which comes in the MML git repo. Compile and run MML from source in order to print units.
+                command = new String[] {
+                    mmlExecutable.getAbsolutePath(),
+                    "run",
+                    "--args=%s --no-startup".formatted(unitFile.getAbsolutePath())
+                };
+            } else {
+                // Start mml normally. "--no-startup" tells MML to exit after the user closes the
+                // print dialog (by printing or cancelling)
+                command = new String[] {
+                    mmlExecutable.getAbsolutePath(),
+                    unitFile.getAbsolutePath(),
+                    "--no-startup"
+                };
+            }
+        } else {
+            if (!mmlExecutable.exists()) {
+                if (autodetected) {
+                    logger.error("Could not auto-detect MegaMekLab! Please configure the path to the MegaMekLab executable in the settings.", "Error printing unit list");
+                } else {
+                    logger.error("%s does not appear to exist! Please configure the path to the MegaMekLab executable in the settings.".formatted(mmlExecutable.getName()), "Error printing unit list");
+                }
+                return null;
+            }
+
+            // The executable is a jarfile, so let's execute it.
+            var javaExecutable = ProcessHandle.current().info().command().orElse("java");
```

Comment:
I'm not a huge fan of this approach; if anything replaces the MML jar file with malicious Java, this will execute it without any warning.

Although, to be fair, the same thing would happen if the MML jar file were replaced with malicious Java and the user just double-clicked it or one of the batch files.

Given that MegaMek is the core program, I'd feel much better about moving low-level printing functionality to MM and then importing it into MML, rather than running MML as a sub-process.

That said, this doesn't seem to expose us to any more security risks, and is much more compact than the full re-write would be.

Reply:
If MegaMekLab.jar is replaced with a compromised version, you're completely screwed anyway. There's no reason to suspect that MegaMek.jar wouldn't also be compromised in such a scenario.